### ISSUE TITLE
Device disconnection on waking up other thread

### ISSUE THREAD

**Issue by NoxZet** [ofTheo/videoInput#44]
I'm using videoInput library for an application that tracks position of a device using multiple cameras. Using it, I got as far as extracting points from input image, but it stopped working once I moved the code to multiple thread.

The main thread creates the capture threads and handles frames passed via vector. This is the initialization that happens in the main thread:
```
videoInput::setComMultiThreaded(true);
videoInput::listDevices(true);
```

In the capture thread, the following code is performed on member object `videoInput cap`
```
cap.setupDevice(deviceId, width, height);
cap.setAutoReconnectOnFreeze(deviceId, true, 1);
```

And then the following code is executed in loop.
```
unsigned char* frame = new unsigned char[width * height * 3]();
cap.isFrameNew(capId);
cap.getPixels(capId, frame, false, true);
```

It works just as intended, but as soon as the data are passed to the main thread, the main thread is notified of the data and woken up with conditional variable, the device is disconnected, resulting in exception on the next iteration of the loop. I got the latest source compiled into a x64 static library.

I absolutely need the application to be multithreaded. Is there something important I am missing to make this work, or have I hit some limitation of this library and need to use something else/make my own DirectShow wrapper?

**Comment by ehardin2** [comment]
I was new to threads but had to do something similar where one thread captured a frame from the camera, did some preprocessing and then stored the frame in a queue.  The second thread would go to the queue for each frame it needed.  One thing that I found was that any memory variable or object that had to be changed or written to had to have the "volatile" keyword used as in
volatile float fCurrentTime = 0;
volatile int iFrame = 0;

The other thing I found was that things had to be locked down when one thread was messing with them so that the other thread would not be able to change things in the middle of an operation.  This was done via std::mutex mutexCamera; as in
mutexCamera.lock();

I don't know if any of this might be causing your problems but it was something that took me awhile to figure out.

Ed

At 08:35 AM 9/9/2018, NoxZet wrote:
